### PR TITLE
Chore avoid failing when sci gets out of sync

### PR DESCRIPTION
**Pull request by Jakub89**
Resolves: #1077

### CODE REVIEW

File: concent_api/core/decorators.py [hunk 1]
```diff
@@ -225,6 +226,11 @@ def wrapper(
             except UnsupportedProtocolVersion as exception:
                 return HttpResponse(exception.error_message, status=404)
 
+            except SCINotSynchronize:
+                json_response = JsonResponse({'error': 'SCI is currently not synchronize'}, status=502)
```
Comment (dybi) [inline concent_api/core/decorators.py]:
```suggestion
                json_response = JsonResponse({'error': 'SCI is currently not synchronized'}, status=502)
```

File: concent_api/core/exceptions.py [hunk 2]
```diff
@@ -83,3 +83,7 @@ class BanksterTimestampError(Exception):
 
 class BanksterTransactionMismatchError(Exception):
     pass
+
+
+class SCINotSynchronize(Exception):
```
Comment (dybi) [inline concent_api/core/exceptions.py]: `SCINotSynchronize` -> `SCINotSynchronized`

File: concent_api/core/payments/backends/sci_backend.py [hunk 3]
```diff
@@ -20,6 +23,26 @@ class TransactionType(Enum):
...
+    def wrapper(*args: Any, **kwargs: Any) -> None:
+        if PaymentInterface().is_synchronized():  # pylint: disable=no-member
+            try:
+                return sci_function(*args, **kwargs)
+            except ValueError as exception:
+                if "There are currently no blocks after" in str(exception):
+                    return []
+                else:
+                    raise
+            except Exception:
```
Comment (dybi) [inline concent_api/core/payments/backends/sci_backend.py]: I think you don't need this...

File: concent_api/core/payments/backends/sci_backend.py [hunk 4]
```diff
@@ -20,6 +23,26 @@ class TransactionType(Enum):
     FORCE = 'force'
 
 
+def handle_exception_if_sci_not_synchronized(sci_function: Callable) -> Callable:
```
Comment (dybi) [inline concent_api/core/payments/backends/sci_backend.py]: `handle_exception_if_sci_not_synchronized` -> `handle_sci_synchronization` or `sci_sync` or sth similar

File: concent_api/core/tests/test_payments_sci_backend.py [hunk 5]
```diff
@@ -280,3 +275,70 @@ def test_that_sci_backend_cover_additional_verification_cost_should_return_trans
             value=self.transaction_value,
             subtask_id=sci_backend._hexencode_uuid(self.task_to_compute.subtask_id),
         )
+
+    def test_that_handle_exception_if_sci_not_synchronized_raise_custom_exception_if_not_sync(self):
+
+        @handle_exception_if_sci_not_synchronized
+        def dummy_handle_exception_if_sci_not_synchronized(_sci_function):
```
Comment (dybi) [inline concent_api/core/tests/test_payments_sci_backend.py]: you don't need `_sci_function` here and `None` in the function call

File: concent_api/core/tests/test_payments_sci_backend.py [hunk 6]
```diff
@@ -280,3 +275,70 @@ def test_that_sci_backend_cover_additional_verification_cost_should_return_trans
...
+                'core.payments.payment_interface.PaymentInterface.__new__',
+                return_value=mock.Mock(
+                    is_synchronized=mock.Mock(
+                        return_value=True,
+                    ),
+                )
+        ):
+            self.assertEqual(dummy_handle_exception_if_sci_not_synchronized(None), [])
+
+    def test_that_handle_exception_if_sci_not_synchronized_raise_not_specific_value_error(self):
```
Comment (dybi) [inline concent_api/core/tests/test_payments_sci_backend.py]: "...raise_not_specific..." -> "...raise**s**_not_specific.."

File: concent_api/core/tests/test_payments_sci_backend.py [hunk 7]
```diff
@@ -280,3 +275,70 @@ def test_that_sci_backend_cover_additional_verification_cost_should_return_trans
...
+                return_value=mock.Mock(
+                    is_synchronized=mock.Mock(
+                        return_value=True,
+                    ),
+                )
+        ):
+            with self.assertRaises(ValueError):
+                dummy_handle_exception_if_sci_not_synchronized(None)
+
+    def test_that_handle_exception_if_sci_not_synchronized_raise_error(self):
```
Comment (dybi) [inline concent_api/core/tests/test_payments_sci_backend.py]: "raise_error" -> "does_not_catch_other_exceptions"?

File: concent_api/core/decorators.py (hunk 1 above)
Comment (Jakub89) [reply concent_api/core/decorators.py]: Done

File: concent_api/core/exceptions.py (hunk 2 above)
Comment (Jakub89) [reply concent_api/core/exceptions.py]: Done

File: concent_api/core/payments/backends/sci_backend.py (hunk 3 above)
Comment (Jakub89) [reply concent_api/core/payments/backends/sci_backend.py]: Done

File: concent_api/core/payments/backends/sci_backend.py (hunk 4 above)
Comment (Jakub89) [reply concent_api/core/payments/backends/sci_backend.py]: Done

File: concent_api/core/tests/test_payments_sci_backend.py (hunk 5 above)
Comment (Jakub89) [reply concent_api/core/tests/test_payments_sci_backend.py]: Done

File: concent_api/core/tests/test_payments_sci_backend.py (hunk 6 above)
Comment (Jakub89) [reply concent_api/core/tests/test_payments_sci_backend.py]: Done

File: concent_api/core/tests/test_payments_sci_backend.py (hunk 7 above)
Comment (Jakub89) [reply concent_api/core/tests/test_payments_sci_backend.py]: Done

File: concent_api/common/tests/test_decorators.py [hunk 8]
```diff
@@ -21,10 +21,11 @@
 from core.decorators import handle_errors_and_responses
 from core.decorators import require_golem_auth_message
 from core.exceptions import Http400
+from core.exceptions import SCINotSynchronized
 from core.tests.utils import ConcentIntegrationTestCase
 from core.tests.utils import parse_iso_date_to_timestamp
 
-(CONCENT_PRIVATE_KEY, CONCENT_PUBLIC_KEY)   = generate_ecc_key_pair()
+(CONCENT_PRIVATE_KEY, CONCENT_PUBLIC_KEY) = generate_ecc_key_pair()
```
Comment (rwrzesien) [inline concent_api/common/tests/test_decorators.py]: Why not using keys that are already available in `ConcentIntegrationTestCase` class?

File: concent_api/core/decorators.py [hunk 9]
```diff
@@ -225,6 +226,11 @@ def wrapper(
             except UnsupportedProtocolVersion as exception:
                 return HttpResponse(exception.error_message, status=404)
 
+            except SCINotSynchronized:
+                json_response = JsonResponse({'error': 'SCI is currently not synchronized'}, status=502)
```
Comment (rwrzesien) [inline concent_api/core/decorators.py]: I think we should use error code for this, as the client might want to handle this situation somehow.

File: concent_api/core/payments/backends/sci_backend.py [hunk 10]
```diff
@@ -20,6 +23,24 @@ class TransactionType(Enum):
...
+        if PaymentInterface().is_synchronized():  # pylint: disable=no-member
+            try:
+                return sci_function(*args, **kwargs)
+            except ValueError as exception:
+                if "There are currently no blocks after" in str(exception):
+                    return []
+                else:
+                    raise
+        else:
+            raise SCINotSynchronized()
```
Comment (rwrzesien) [inline concent_api/core/payments/backends/sci_backend.py]: You may add `'SCI is currently not synchronized'` message to the exception body.

File: concent_api/core/decorators.py (hunk 9 above)
Comment (Jakub89) [reply concent_api/core/decorators.py]: Done

File: concent_api/core/payments/backends/sci_backend.py (hunk 10 above)
Comment (Jakub89) [reply concent_api/core/payments/backends/sci_backend.py]: Done

**Pull request merged**